File: NeetCode/Intervals/Meeting_Rooms_II/meeting__rooms__i_i.py
from typing import Optional, List


# Definition of Interval:
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end


class Solution:
    def minMeetingRooms(self, intervals: List[Interval]) -> int:
        if not intervals:
            return 0

        startTimes = [i.start for i in intervals]
        endTimes = [e.end for e in intervals]
        startTimes.sort()
        endTimes.sort()

        sPtr, ePtr, count, maxCount = 0, 0, 0, 0
        size = len(intervals)

        while sPtr < size and ePtr < size:
            if startTimes[sPtr] < endTimes[ePtr]:
                sPtr += 1
                count += 1
            else:
                ePtr += 1
                count -= 1
            maxCount = max(maxCount, count)
        return maxCount


if __name__ == "__main__":
    sol = Solution()
    intervals = [(0, 40), (5, 10), (15, 20)]
    intervals = [Interval(i[0], i[1]) for i in intervals]
    print(sol.minMeetingRooms(intervals))

    print("Running Solution...")

File: NeetCode/Intervals/Meeting_Rooms_II/Meeting_Rooms_II.cpp
// Meeting_Rooms_II.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Interval
{
public:
    int start, end;
    Interval(int start, int end)
    {
        this->start = start;
        this->end = end;
    }
};

class Solution
{
public:
    int minMeetingRooms(vector<vector<int>> &intervals)
    {
        {
            if (intervals.size() == 0)
            {
                return 0;
            }
            vector<vector<int>> startTimes(intervals);
            vector<vector<int>> endTimes(intervals);
            sort(startTimes.begin(), startTimes.end(), [](const vector<int> &a, const vector<int> &b)
                 { return a[0] < b[0]; });
            sort(endTimes.begin(), endTimes.end(), [](const vector<int> &a, const vector<int> &b)
                 { return a[1] < b[1]; });

            int sPtr(0), ePtr(0), count(0), maxCount(0);

            while (sPtr < intervals.size() && ePtr < intervals.size())
            {
                if (startTimes[sPtr][0] < endTimes[ePtr][1])
                {
                    sPtr++;
                    count++;
                }
                else
                {
                    ePtr++;
                    count--;
                }
                maxCount = max(maxCount, count);
            }
            return maxCount;
        }
    }
};

/*
class Solution
{
public:
    int minMeetingRooms(vector<Interval> &intervals)
    {
        if(intervals.size() == 0){
            return 0;
        }
        vector<Interval> startTimes(intervals);
        vector<Interval> endTimes(intervals);
        sort(startTimes.begin(), startTimes.end(), [](const Interval &a, const Interval &b){
            return a.start < b.start;
        });
        sort(endTimes.begin(), endTimes.end(), [](const Interval &a, const Interval &b){
            return a.end < b.end;
        });

        int sPtr(0), ePtr(0), count(0), maxCount(0);

        while( sPtr < intervals.size() && ePtr < intervals.size()){
            if(startTimes[sPtr].start < endTimes[ePtr].end){
                sPtr++;
                count++;
            }else{
                ePtr++;
                count--;
            }
            maxCount = max(maxCount, count);
        }
        return maxCount;
    }
};

*/